### ISSUE TITLE
[AUTOMATION] Bump axe-core library to 4.1.1 in our CI/CD pipeline to stay in sync with browser plugin

### ISSUE THREAD

**Issue by 1Copenut** [department-of-veterans-affairs/va.gov-team#16544]
## Potential Blocker
We should not merge this work until https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/357 has been resolved. The Design System progress bar has been identified as throwing an axe violation using the latest ruleset. A proposed fix was brought up to the Design System Council on 12/11, and the PM notified of the issue ticket.

## Issue Description
Requesting to bump the axe-core library from the current `v3.5.4` to the latest stable release `v4.1.1`. This library runs all of our automated accessibility checks in the Cypress e2e tests and static content checks during the CI/CD build cycle.

The [front-end browser plugin](https://deque.com/axe) we use for manual testing in the VSP Collaboration Cycle is using the current 4.1.1 ruleset, so we'd like to keep the CI ruleset in sync.

---
## Tasks
- [x] package.json has to be bumped to use the current version `$ npm update axe-core`
- [x] Cypress axeCheck helper method will need to be revised to accept a custom configuration option
- [x] All e2e tests should be run locally to ensure no breaks before we attempt to merge the branch
- [x] Update cypress-axe to the latest version (0.12.0 as of writing)
- [x] Disable the axe color-contrast rule to improve testing performance and reduce axe false positives
- [x] Extend the axe helper method to selectively re-enable the color contrast rule
- [x] Extend the axe helper method to selectively enable the Best Practices ruleset
- [x] axe documentation is updated to reflect the new functionality
- [x] [cypress-axe](https://www.npmjs.com/package/cypress-axe) has been updated to the latest version

## Acceptance Criteria
- [x] All e2e tests and all static content axe tests must pass in the CI/CD pipeline
- [x] Devs will need the ability to pass a custom config object. This must allow developers to ignore individual rules as we often get false positive color contrast violations in the CI server.
- [x] Color contrast issues are ignored on all Cypress tests
- [x] Color contrast tests can be selectively re-enabled

## Relevant Files and Issues
* [axe-core 4.1.1 release notes](https://github.com/dequelabs/axe-core/releases/tag/v4.1.1)
* [Legacy Nightwatch axeCheck helper](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/testing/e2e/nightwatch-commands/axeCheck.js) -- focus on lines 5, 30, and 65.
  * We are removing certain elements and rules from our testing in this legacy test suite. We want to make sure we don't run into color contrast errors in CI, without ignoring the rule completely if possible.
* [Cypress axeCheck helper](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/testing/e2e/cypress/support/commands/axeCheck.js)
* [Issue 13647 with some related changes to the Cypress helper](https://github.com/department-of-veterans-affairs/va.gov-team/issues/13647) 
* [Issue 17257 to turn off the color contrast rule and re-enable selectively](https://github.com/department-of-veterans-affairs/va.gov-team/issues/17257)

---
## How to configure this issue
- [ ] **Attached to a Milestone** (when will this be completed?)
- [ ] **Attached to an Epic** (what body of work is this a part of?)
- [x] **Labeled with Team** (`product support`, `analytics-insights`, `operations`, `service-design`, `tools-be`, `tools-fe`)
- [x] **Labeled with Practice Area** (`backend`, `frontend`, `devops`, `design`, `research`, `product`, `ia`, `qa`, `analytics`, `contact center`, `research`, `accessibility`, `content`)
- [x] **Labeled with Type** (`bug`, `request`, `discovery`, `documentation`, etc.)

**Comment by holdenhinkle** [comment]
List of applications that call axeCheck():
- vaos
- personalization
- healthcare
- gi
- facility locator
- disability benefits
- debt-letters

Upgraded `axe-core` to v4.1.1 and manually ran each test that calls cy.axeCheck() in the Cypress Test Runner (not the headless version). All tests are passing. Well, there are no changes (one test in `src/applications/facility-locator/tests/e2e/mobile.cypress.spec.js` still fails, but that doesn't have to do with the `axe-core` module).

I'm currently re-running all tests in headless mode (`yarn cy:run`)...

**Comment by holdenhinkle** [comment]
This ticket is completed.

It looks like we're going to hold the `axe-core` bump to v4.1.1 until January.

This task--Cypress axeCheck helper method will need to be revised to accept a custom configuration option--per my conversation with Trevor, is not applicable.

This acceptance criterion--Devs will need the ability to pass a custom config object. This must allow developers to ignore individual rules as we often get false positive color contrast violations in the CI server--is also not applicable.

When all Cypress tests are run in headless mode, 5-7 tests always seem to fail. However, when each failed test is run individually in either test mode or in the visual test runner, they pass.

We don't know if other tests that rely on `axe-core` will pass or not in CI. Trevor suggested that we start a discussion on VFS Engineers about this. He mentioned the following people:
-Michael Fleet
- Dan Shank
- Chris Valarida
- Nick Sullivan

If tests start failing, preventing merges, we want to be able to handle it quickly.

**Comment by JoeTice** [comment]
Moving to future initiatives till after holiday code freeze, will plan to execute this on Jan 4 2021

**Comment by 1Copenut** [comment]
This issue has two dependencies before it can be merged. Should be considered blocked until they are resolved:

* https://github.com/department-of-veterans-affairs/va.gov-team/issues/17257
* https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/357

**Comment by pjhill** [comment]
For [the false positive on contrast ratio issue](https://github.com/department-of-veterans-affairs/va.gov-team/issues/17257), we discussed a potential way forward. @1Copenut will discuss turning off detection of contrast issues through automated e2e tests given our coverage through other means (a11y checklist manual scans and react component unit tests) with the VA's 508 office.

**Comment by pjhill** [comment]
## Dependencies
- [ ] Configuration object
  * Complexity: 3
- [ ] Design System issues (progress bar & loading indicator)
  * Complexity Progress Bar: 3
  * Complexity Loading Indicator: 3
- [ ] VAMC
  *

**Comment by 1Copenut** [comment]
@JoeTice All work is completed on this issue. The last item is documentation, in PR now: https://github.com/department-of-veterans-affairs/va.gov-team/pull/20058.

Moving this to close.
